test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc

```diff
@@ -927,6 +940,30 @@ TEST_P(ProxyProtocolTest, PartialRead) {
   disconnect();
 }
 
+TEST_P(ProxyProtocolTest, PartialV1ReadWithAllowNoProxyProtocol) {
+
+  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config;
+  proto_config.set_allow_requests_without_proxy_protocol(true);
+  connect(true, &proto_config);
+
+  write("PROXY TCP4");
```

Comment:
We should test the case of the first write is not full v1 protocol signature, like just one char 'P'

Reply:
I'm not sure I understand how this is possible to support with the given use case? We need to peek, the second `recv` doesn't use any flags so it will consume data we might want to forward downstream off the socket.

We could put `recv` with `MSG_PEEK` in a loop until we can see enough bytes to make the determination on whether we have proxy protocol or not (similar to your timeout proposal), but this has bad implications for the downstream request as well.

Reply:
Addressed in https://github.com/envoyproxy/envoy/pull/18951/commits/6a0d49bc4b1bfaca07ac3132243f2d42dfff5fb7

Reply:
I'm actually not sure this is possible simultaneously with the fragmented/partial protocol without the possibility for false positive, e.g.
```c++
TEST_P(ProxyProtocolTest, TinyTODOPartialV1ReadWithAllowNoProxyProtocol) {

  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config;
  proto_config.set_allow_requests_without_proxy_protocol(true);
  connect(true, &proto_config);

  write("P");

  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);

  write("BOGUS");

  expectData("PBOGUS"); // we will consume and get BOGUS not PBOGUS

  disconnect();
}
```
this is why in the initial implementation I required that the initial read from `recv` have enough bytes to make the determination on the v1/v2 proxy header.

We can leave the PR as it is with the understanding that there could be false positives on identifying proxy protocol, with the understanding that this is quite rare and will not likely happen in practice. In exchange, we get support for fragmented/partial reads. Thoughts?

Reply:
I'm not following. How would the filter consume the `P` only?

Reply:
Running the test above, we do `recv` with `MSG_PEEK` and just get `P` [here](https://github.com/envoyproxy/envoy/blob/v1.20.0/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L423).

Then before we continue the loop, we read that byte for real (no `MSG_PEEK`) [here](https://github.com/envoyproxy/envoy/blob/v1.20.0/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L528). Thus when we do our second `recv` and we get `BOGUS` we have already consumed `P`. We correctly identify that the request is not proxy protocol, but the request forwarded upstream has a missing byte(s) (in this case, just `P`).

The only way I see to avoid false positives identifying the header and always send the correct response up the filter chain is to require that the initial `MSG_PEEK` has enough bytes to detect v1/v2 header (at most 16, which seems reasonable for this opt-in only codepath)

Reply:
Oh, I see.

I think you need to fix this case. You may have to restructure the code such that no bytes are read (without MSG_PEEK) until we've decided 100% whether this is a proxy header or not.

I think it's probably good enough to look for either `PROXY_PROTO_V1_SIGNATURE` or `PROXY_PROTO_V2_SIGNATURE`. If the connections starts with either of those, but then doesn't have a valid full proxy protocol header, abort the connection as we currently do. If the connection starts with bytes that match neither of those signatures, and the new option is enabled, don't `read()` any bytes and continue the filter chain. Will that work?

Reply:
Addressed in https://github.com/envoyproxy/envoy/pull/18951/commits/6f19883279fa64bbcc107f0a7d21da1cedc71da0

Reply:
Sorry for I didn't describe it clearly.

The case what I'm thinking is:

TEST_P(ProxyProtocolTest, partialv1Signature) {

  envoy::extensions::filters::listener::proxy_protocol::v3::ProxyProtocol proto_config;
  proto_config.set_allow_requests_without_proxy_protocol(true);
  connect(true, &proto_config);

  write("P");

  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
  write("ROXY TCP4");
  write(" 254.254.2");
  write("54.254 1.2");
  write(".3.4 65535");
  write(" 1234\r\n...");

  expectData("...");

  EXPECT_EQ(server_connection_->connectionInfoProvider().remoteAddress()->ip()->addressAsString(),
            "254.254.254.254");
  EXPECT_TRUE(server_connection_->connectionInfoProvider().localAddressRestored());

  disconnect();
}